enhancements/kube-apiserver/api-webhook-supportability.md

```diff
@@ -0,0 +1,378 @@
+  tokens or secrets. The webhook will see these in plain-text and hence become a security risk for
+  the cluster.
+- the **availability** of a cluster because API requests that fail can render resource-agnostic
+  controllers like garbage collection, namespace controller or resource quota disfunctional for all
+  resources in a cluster, not only those the webhook is about.
+
+### Goals
+
+- allow the customer, support and engineering in escalations to notice misbehaving webhooks
+- make the customer aware about risks of installing (multiple) add-ons that hook deeply into core
+  components
+- make Red Hat teams aware early about the risk of hooking into core resources and make them design
+  their architecture with minimal impact
+- establish an Red Hat approval process for webhooks hooking into core resources
+- establish a culture of "webhook impact budget" in the development process.
+
+### Non-Goals
+
+- tainting a cluster with effect on the support status
+- forbidding webhooks for OpenShift add-ons
+- out-of-scope: proposing alerts for kube-controller-manager, which can stopped from doing its work
+  (GC, namespace controller, quota) with unstable webhooks (especially conversion webhooks)
+- out-of-scope: creating alerts for aggregated apiservers
+
+## Proposal
+
+### User Stories
+
+1. As a cluster-admin I want to be warned if core resource webhooks risk the cluster health.
+2. As a cluster-admin I want to be warned if webhooks are slow and risk the user-experience or even
+   the availability of non-core resources (CRDs).
+3. As a cluster-admin I want to be notified and acknowledge the risk of core resource webhooks that
+   are being installed.
+4. As a development organization, we want to develop architectures that do not harm or minimally
+   harm the stability, performance, availability and security of OpenShift.
+5. As a development organization, we don't want to stop innovation and hence allow a decentralized
+   development process of OpenShift add-ons, but establish a culture where best-practices are
+   communicated (e.g. in architecture reviews) and applied.
+
+### Metrics
+
+Upstream kubernetes provides the following webhook specific metrics:
+
+- histogram `webhook_admission_duration_seconds(name, type, operation, rejected)`
+- histogram `webhook_admission_duration_seconds_total(type, operation, rejected)`
+- counter `webhook_rejection_count(name, type, operation, error_type, rejection_code)`
+
+### Critical Alerts
+
+We will add the following critical alerts:
+
+- alert about latency per webhook > 0.5s (0.2s for critical resources)
+- alert about total webhook latency per resource > 1s (0.2 for critical resources)
+- alert about virtual resources (SubjectAccessReview, ...) webhooks
+- alert about security sensitive webhooks (TokenReviews, oauth tokens+clients)
+- alert about failure count > 1% (both conversion and admission).
+- degraded if webhooks's service does not exist, or has no endpoints
```

Comment:
will this be a degraded condition on kas-o operator?

Reply:
yes